Proofreading chapter 2 (#4487)
kptdev/kpt: documentation/content/en/book/02-concepts/_index.md

---
title: "Chapter 2: Concepts"
linkTitle: "Chapter 2: Concepts"

> description: This chapter describes what kpt is and what the main concepts are behind kpt.

Reviewer: ```suggestion
description: This chapter describes what kpt is and what the main concepts behind it are.
```

Reviewer: i just thought the description read weird to me.

Reviewer 2: I've reworked even your suggestion in [a102120](https://github.com/kptdev/kpt/pull/4487/commits/a1021204757564fa6d2de6b8ab61012b229dafc2)

## What is kpt?

kpt stands for Kubernetes Package Transformation. It supports the management of configuration as data (CaD).

> *Configuration as data* is an approach to the management of configurations, including the configuration of infrastructure, policy, services, applications, and so on, which comprises the following actions:

Reviewer: ```suggestion
*Configuration as data* is an approach to the management of configurations, including the configuration of infrastructure, policy, services, applications, among others, which comprises the following actions:
```
and so on doesn't sound very professional

However, kpt optimizes for in-place configuration transformation rather than out-of-place transformation. 

Validation goes hand-in-hand with customization. KRM functions can be used to automate both mutation and validation of resources, similarly to 
[Kubernetes admission control](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/). 

A kpt package is a bundle of configuration _data_. It is represented as a directory tree containing the KRM resources using YAML as the file format.

A package is explicitly declared using a file named `Kptfile`. This file contains a KRM resource of kind `Kptfile`. The Kptfile contains metadata about the package and is simply a regular resource in the YAML format.

Just as directories can be nested, a package can contain another package. This is called a _subpackage_.

### Kptfile annotations

The Kptfile supports annotations that control package-level behavior:

- **`kpt.dev/bfs-rendering`**: When set to `"true"`, this annotation renders the package hierarchy in breadth-first order, instead of the default depth-first
  post-order.

> - **`kpt.dev/save-on-render-failure`**: When set to `"true"`, this annotation saves partially rendered resources to disk, even when rendering fails, instead of reverting all changes. This is particularly useful for debugging render failures and is essential for programmatic package rendering scenarios, where preserving partial progress is valuable.

Reviewer: as we mention subpackages just above this the annotations here should probably also clarify if they operate and the subpackage level when defined in its parent package or if its isolated on a per package basis etc

Reviewer 2: Correct, the only problem that I do not know the answer :)
Will ask on Slack.

Reviewer 2: Fixed in [342a52d](https://github.com/kptdev/kpt/pull/4487/commits/342a52da3050484c918ac1d57ac017cd9028c985)

This _package hierarchy_ contains two packages:

1. `wordpress`: This is the top-level package in the hierarchy. It is declared using `wordpress/Kptfile`. This package contains two subdirectories.
`wordpress/deployment` is a regular directory used for organizing resources that belong to the `wordpress` package itself. The `wordpress` package contains
three direct resources in three files:
  - `service.yaml`
  - `deployment/deployment.yaml`

>   - `deployment/volume.yaml`

Author: This appears to have been rendered correctly.

2. `wordpress/mysql`: This is a subpackage of the `wordpress` package, since it contains a `Kptfile`. This package contains three resources in the 
`wordpress/mysql/deployment.yaml` file.

> 

Author: This appears to have been rendered correctly.

- **Create**: Initialize a package using the `kpt pkg init` command. This command creates the directory if it does not exist.

Let us suppose that you have rendered the package, and would like to deploy it to a cluster. The workflow may look like this:

![img](/images/lifecycle/flow3.svg)

- **Initialize**: This is a one-time process using the `kpt live init` command.
- **Preview**: Use the `kpt live apply --dry-run` command.
- **Apply**: Use the `kpt live apply` command.
- **Observe**: Use the `kpt live status` command.

First, use the kpt `live apply --dry-run` command to validate the resources in your package and verify that the expected resources will be applied and pruned. If the preview looks good, then apply the package, using the `kpt live apply` command. Afterwards, you may observe the status of the package on the cluster.

Typically, it is best to store the package in Git:

![img](/images/lifecycle/flow4.svg)

- **Publish**: Use the `git commit` command.

The publishing flow is orthogonal to the deployment flow. This allows you to act as a publisher of an upstream package, even though you may not deploy the package personally.

## Functions

> A Kubernetes Resource Model (KRM) function (formerly called a _kpt_ function_) is a containerized program that can perform create, read, update, and delete (CRUD) operations on KRM resources stored on the local filesystem. kpt functions are the extensible mechanism to automate the mutation and validation of KRM resources. The following are some example use cases:

Author: Fixed in [cec97b1](https://github.com/kptdev/kpt/pull/4487/commits/cec97b1827b4f5d6e0a22bd96dc7f608cefe3214)

`kpt fn` and `kpt live`) use the package from the local filesystem, rather than the remote Git repository. It can be thought of as the _vendoring_ used by tooling for some programming languages. The main difference is that kpt is designed to enable you to modify the vendored package on the local filesystem, and then update the package by merging the local and upstream changes.

There is one scenario where a Kptfile is implicit: you can use kpt to fetch any Git directory containing KRM resources, even if the directory does not contain a `Kptfile`. Effectively, you are telling kpt to treat the Git directory as a package. kpt automatically creates the `Kptfile` on the local filesystem to keep track of the upstream repository. This means that kpt is compatible with a
large corpus of existing Kubernetes configurations currently stored on Git.

For example, `spark` is essentially a vanilla directory of KRM:

```shell
kpt pkg get https://github.com/kubernetes/examples/tree/master/_archived/spark
```

<<<<<<< HEAD
Details of how to work with packages are set out in [Chapter 3](../03-packages).

> =======

Reviewer 3: Merge conflict need to be resolved.

To astute readers, this model will sound familiar: the functions are the client-side analog to the Kubernetes controllers:

|                  | Client-side              | Server-side       |
| ---------------- | ------------------------ | ----------------- |
| **Orchestrator** | kpt                      | Kubernetes        |
| **Data**         | YAML files on filesystem | resources on etcd |
| **Programs**     | functions                | controllers       |

Just as the Kubernetes system orchestrates the server-side containers, the kpt CLI orchestrates the client-side containers operating on the configuration. By standardizing the input and output of the function containers, and how the containers are executed, kpt can provide the following guarantees:

- The functions are interoperable.
- The functions can be chained together.
- The functions are hermetic. For correctness, security and speed, it is desirable to be able to run functions hermetically without any privileges, thereby preventing out-of-band access to the host filesystem and networking.

We will discuss the KRM Functions Specification Standard in detail in

> <<<<<<< HEAD

Reviewer 3: Merge conflict need to be resolved.

- `input items`: This is the input list of the KRM resources on which to operate.
- `output items`: This is the output list obtained from adding, removing, or modifying items in the input.
- `functionConfig`: This is an optional meta resource containing the arguments to this invocation of the function.
- `results`: This is an optional meta resource emitted by the function for observability and debugging purposes.

Functions can be chained together in a pipeline, as illustrated here:

![img](/images/pipeline.svg)

There are two different CLI commands that execute the functions corresponding to two fundamentally different approaches:

- `kpt fn render`: This command executes the pipeline of functions declared in the package and its subpackages. This is a declarative way to run the functions.
- `kpt fn eval`: This command executes a given function on the package. The image to run and the `functionConfig` are specified as a CLI argument. This is an imperative way to run functions. Since the function is provided explicitly by the user, an imperative invocation can be more privileged and low-level than a declarative invocation. For example, it can have access to the host system.

> <<<<<<< HEAD

Reviewer 3: Merge conflict need to be resolved.
